### PR TITLE
Fix Kimi-K2.5 accuracy when Aiter MLA FP8 PS + CUDA graphs are used

### PR DESCRIPTION
**Pull request by xaguilar-amd**
## Problem

When running MLA-based models (e.g., Kimi-K2.5-MXFP4) with `--kv-cache-dtype fp8` and FULL CUDA graph mode, the model outputs garbage/gibberish text. The issue does **not** occur with:
- `--enforce-eager` (no CUDA graphs)
- `--compilation-config '{"cudagraph_mode": "NONE"}'`
- `--compilation-config '{"cudagraph_mode": "PIECEWISE"}'` 
- Non-FP8 KV cache (e.g., `--kv-cache-dtype auto`)

There are **two independent bugs**:

1. **KV cache corruption during warmup** — `concat_and_cache_mla` writes garbage FP8 data to slot 0 during dummy runs because `slot_mapping` is uninitialized (zeros).
2. **Missing FP8 dtype in `get_mla_metadata_v1`** — the persistent kernel's work scheduling metadata is computed without `dtype_q` and `dtype_kv`, causing it to assume BF16 data layout. With FP8 KV cache, this produces incorrect memory access patterns, so the kernel reads from wrong KV cache positions.

---

### Root Cause

During the CUDA graph warmup phase in `_dummy_run()`, vLLM executes dummy forward passes with `force_attention=True` to warm up the attention kernels. This triggers the full attention pipeline including **KV cache writes** via `concat_and_cache_mla`.

The `slot_mapping` buffer used for these writes is read from `self.input_batch.block_table[gid].slot_mapping.gpu`, which is a `CpuGpuBuffer` initialized to zeros via `torch.zeros()`. Since no real requests have been scheduled at warmup time, the buffer has never been populated by the scheduler.

With `slot_mapping` = all zeros:
1. `concat_and_cache_mla` writes FP8-quantized garbage KV data to **slot 0** of the KV cache for **every layer** and **every dummy token**
2. This corrupts the KV cache state before any real inference begins
3. The persistent MLA decode kernel (`_ps` variant) reads from the corrupted slot 0 via its work scheduling metadata for padded batch entries, producing garbage attention output

### Why PIECEWISE mode is unaffected

The warmup `force_attention` flag is set based on:
```python
force_attention = cudagraph_runtime_mode == CUDAGraphMode.FULL
```

In PIECEWISE mode, the warmup dummy runs use `CUDAGraphMode.NONE`, so `force_attention=False`, and the attention path (including KV cache writes) is skipped entirely. Only FULL mode triggers the problematic warmup path.

### Root Cause

The `get_mla_metadata_v1()` call in `rocm_aiter_mla.py` (`_build_decode` method) is missing three parameters:

- `dtype_q` — the query data type (should be FP8 when KV cache is FP8)
- `dtype_kv` — the KV cache data type (should be FP8 when KV cache is FP8)
- `max_split_per_batch` — controls the reduction splitting strategy (should be 16)

Without these, `get_mla_metadata_v1` computes work scheduling metadata assuming BF16 data layout. Since FP8 and BF16 have different element sizes (1 byte vs 2 bytes), the resulting work metadata contains incorrect memory offsets and access patterns. The persistent kernel then reads from wrong positions in the KV cache, producing garbage attention output even when the KV cache data itself is correct.

Additionally, `get_mla_metadata_info_v1()` (which computes buffer *sizes*) receives `q_dtype = self.decode_attn_out_dtype` (BF16), while `get_mla_metadata_v1()` needs `dtype_q = FP8`. This dtype mismatch means the persistent buffers (`work_meta_data`, `reduce_indptr`, etc.) are sized for a BF16-query work plan, but the metadata computation tries to write an FP8-query work plan. This can cause the metadata function to produce a constrained or suboptimal scheduling plan.

## Test Plan

- [ x] Verify Kimi-K2.5-MXFP4 with `--kv-cache-dtype fp8 -tp 4` produces correct output (FULL CG mode, default)
- [ x] Verify PIECEWISE mode still works
- [ x] Verify `--enforce-eager` still works
- [ x] Run `lm_eval` accuracy benchmark to confirm no regression

Accuracy with lm_eval and gsm8k before the fix 0.

After the fix:
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9287|±  |0.0071|
|     |       |strict-match    |     5|exact_match|↑  |0.9295|±  |0.0071|

### CODE REVIEW

**Comment by Rohan138** [comment]
Hi @xaguilar-amd can you rebase against main again? We'll get it reviewed+merged right after

**Comment by xaguilar-amd** [comment]
Closing this PR.

After re-testing today against the latest upstream `main` + latest AITER, the Kimi-K2.5-MXFP4 accuracy regression with `--kv-cache-dtype fp8` and FULL CUDA graphs no longer reproduces. Tracing what changed since I opened this PR, the fix landed in a newer PR #39444 ("[Bugfix] Fix V1 dummy run writing NaN to KV cache null block", merged 2026-04-10).

That PR is functionally identical to "Bug 1" in this PR's description: it fills `slot_mapping` with `-1` (== `PAD_SLOT_ID`) in `_dummy_run` after `_get_slot_mappings`, preventing `concat_and_cache_mla` from writing FP8-quantized garbage into block 0 (the null block) during warmup. That null-block corruption was what the persistent MLA decode kernel was reading back as "garbage attention output" — same root cause, same fix, just discovered independently on DeepSeek R1 NVFP4 with DP+EP. That's also why the diff in this PR shrank to only `rocm_aiter_mla.py` after my April 10 rebase — by then the slot_mapping hunk was already a no-op against `main`.

If passing the dtypes through is still desirable as a hardening / consistency fix (so `get_mla_metadata_info_v1` and `get_mla_metadata_v1` agree (as this PR does), and so configurations that actually hit the non-16-head FP8 branches get the right scheduling), I'm happy to re-open a smaller, re-scoped PR for just that. But the accuracy issue this PR was opened for is resolved by #39444, so closing.

Thanks @elvircrn for the parallel fix, and thanks @dllehr-amd / @Rohan138 / @gshtras for the reviews here.

**Comment by dllehr-amd** [comment]
Thanks for confirming and closing @xaguilar-amd !
